src/geometry.h

```diff
@@ -0,0 +1,1104 @@
+template <typename T> class Vector2 {
+public:
+    Vector2() = default;
+    Vector2(T x, T y) : x{x}, y{y} { assert(!contains_nan()); }
+
+    template <typename U>
+    explicit Vector2(const Vector2<U>& v)
+        : x{static_cast<T>(v.x)}, y{static_cast<T>(v.y)} {
+        assert(!contains_nan());
+    }
+
+    template <typename U> Vector2<U> as() const { return Vector2<U>(*this); }
+
+    bool contains_nan() const { return std::isnan(x) || std::isnan(y); }
+
+    T operator[](Axis2 ax) const { return ax == Axis2::X ? x : y; }
+
+    T& operator[](Axis2 ax) { return ax == Axis2::X ? x : y; }
+
+    T operator[](size_t i) const {
+        assert(i < 2);
+        return i == 0 ? x : y;
+    }
+
+    T& operator[](size_t i) {
+        assert(i < 2);
+        return i == 0 ? x : y;
+    }
+
+    bool operator==(const Vector2& v) const { return x == v.x && y == v.y; }
+
+    bool operator!=(const Vector2& v) const { return !(*this == v); }
+
+    Vector2<T> operator+(const Vector2<T>& v) const {
+        return {x + v.x, y + v.y};
+    }
+
+    Vector2<T>& operator+=(const Vector2<T>& v) {
+        x += v.x;
+        y += v.y;
+        return *this;
+    }
+
+    Vector2<T> operator-(const Vector2<T>& v) const {
+        return {x - v.x, y - v.y};
+    }
+
+    Vector2<T>& operator-=(const Vector2<T>& v) {
+        x -= v.x;
+        y -= v.y;
+        return *this;
+    }
+
+    Vector2<T> operator*(T s) const { return {s * x, s * y}; }
+
+    Vector2<T>& operator*=(T s) {
+        x *= s;
+        y *= s;
+        return *this;
+    }
+
+    Vector2<T> operator/(T s) const {
+        assert(s != 0);
+        return (1 / s) * *this;
+    }
+
+    Vector2<T>& operator/=(T s) {
+        assert(s != 0);
+        return *this *= 1 / s;
+    }
+
+    Vector2<T> operator-() const { return {-x, -y}; }
+
+    T length_squared() const { return x * x + y * y; }
+    T length() const { return std::sqrt(length_squared()); }
+
+    std::string to_string() const {
+        return "[" + std::to_string(x) + ", " + std::to_string(y) + "]";
+    }
+
+public:
+    T x = 0;
+    T y = 0;
+};
+
+using Vector2f = Vector2<float>;
+using Vector2i = Vector2<int>;
+
+template <typename T> Vector2<T> operator*(T s, const Vector2<T>& v) {
+    return {s * v.x, s * v.y};
+}
+
+template <typename T> Vector2<T> abs(const Vector2<T>& v) {
+    return {std::abs(v.x), std::abs(v.y)};
+}
+
+template <typename T> T dot(const Vector2<T>& v1, const Vector2<T>& v2) {
+    return v1.x * v2.x + v1.y * v2.y;
+}
+
+template <typename T> T absdot(const Vector2<T>& v1, const Vector2<T>& v2) {
+    return std::abs(dot(v1, v2));
+}
+
+template <typename T> Vector2<T> normalize(const Vector2<T>& v) {
+    return v / v.length();
+}
+
+template <typename T> T min_component(const Vector2<T>& v) {
+    return std::min(v.x, v.y);
+}
+
+template <typename T> T max_component(const Vector2<T>& v) {
+    return std::max(v.x, v.y);
+}
+
+template <typename T> int max_dimension(const Vector2<T>& v) {
```

Comment:
Not really different. Do you have a better name? This function comes from prbt and is not used anywhere right now.